autorest/tracing/tracing.go

```diff
@@ -0,0 +1,124 @@
+package tracing
+
+// Copyright 2018 Microsoft Corporation
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+import (
+	"context"
+	"net/http"
+	"os"
+
+	"contrib.go.opencensus.io/exporter/ocagent"
+	"go.opencensus.io/stats/view"
+
+	"go.opencensus.io/trace"
+
+	"go.opencensus.io/plugin/ochttp"
+	"go.opencensus.io/plugin/ochttp/propagation/tracecontext"
+)
+
+var (
+	// Transport is the default tracing RoundTripper. If tracing is disabled
+	// the default RoundTripper is used which provides no instrumentation.
+	Transport = http.DefaultTransport
+
+	// Enabled is the flag for marking if tracing is enabled.
+	Enabled = false
+
+	// Sampler is the tracing sampler. If tracing is disabled it will never sample. Otherwise
+	// it will be using the parent sampler or the default.
+	sampler = trace.NeverSample()
+)
+
+func init() {
+	enableFromEnv()
+}
+
+func enableFromEnv() {
+	_, ok := os.LookupEnv("AZURE_SDK_TRACING_ENABELD")
+	if ok {
+		agentEndpoint, ok := os.LookupEnv("OCAGENT_TRACE_EXPORTER_ENDPOINT")
+
+		if ok {
+			EnableWithAIForwarding(agentEndpoint)
+		} else {
+			Enable()
+		}
+	}
+}
+
+// Enable will start instrumentation for metrics and traces.
+func Enable() (err error) {
+	Enabled = true
+	sampler = nil
+	Transport = &ochttp.Transport{Propagation: &tracecontext.HTTPFormat{}}
```

Comment:
note we updated for the W3C's Trace Context for propagating across servers instead of the B3 default implementation. Since Trace Context is on its way of becoming the standard but still a draft a little torn here. @joshgav regarding the ARM tracing, for immediate practical effects using the Trace Context means having only 1 header with any custom key-value pairs as opposed to some predefined headers for the B3 propagation .